nixos/modules/config/zram.nix

```diff
@@ -92,6 +92,46 @@ in
           as there's no gain from keeping them in RAM.
         '';
       };
+
+      use-optimal-sysctls = lib.mkOption {
+        default = true;
+        type = lib.types.bool;
+        description = ''
+          Applies some sysctls to optimize the virtual memory subsystem
+          for zramSwap.
+
+          - `vm.swappiness = 150` increase swapping aka compression to be
+            able to cache more file page data
+          - `vm.watermark_boost_factor = 0` watermark boosting can cause unpredictable stalls as seen here:
+            https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1861359
+          - `vm.watermark_scale_factor = 125` initiate kswapd much earlier
+            as zram will also apply pressure when requesting memory for
+            compressed swap
+          - `vm.page-cluster = 0` disables page prefetching
+
+          This option when enabled, sets these sysctls when you also
+          enable the zramSwap module.
+
+          Defaults to `true`.
+        '';
+      };
+
+      disable-zswap = lib.mkOption {
+        default = true;
+        type = lib.types.bool;
+        description = ''
+          When using zram-backed swap, it is unwise to have zswap enabled.
```

Comment:
But other swap devices can still exist. Is this also considered?

Reply:
No. If you want to have a physically backed swap then you can only:
- configure a backing device for the zram (there is a config for that in zramSwap)
- don't use zramSwap at all, have a swapfile or swap partition and have zswap turned on (this is what zswap was made for, that is its intended purpose and this is the ideal setup)

Combining zramswap and physical swap is not a great idea. Data that is really at the far end of the LRU (least recently used list) gets put into zramswap first because zramswap is mounted with a higher priority. And then if pressure gets higher, the items further up in the LRU list are swapped onto disk. So you have very likely never touched pages in zramswap while the more frequently used pages are paged to disk. Not ideal.

Reply:
I never really checked if NixOS defaults to zswap on by default. This is determined by a kconfig option.

Reply:
In regards to zswap, there is a wonderful line I would like to quote from Archwiki:
> The difference compared to zram is that zswap works in conjunction with a swap device while zram with created swap on top of it is a swap device in RAM that does not require a backing swap device.

Maybe I should adjust the description of the zramSwap module too to make this more clear.

Reply:
A last note: if you disable zswap and you use a on-disk swap, the outcome is not terrible, you just get some more IO done to the disk.

Reply:
@Aleksanaa did that answer your questions?

Reply:
Maybe this should be a warning instead? Also, I don't think our kernel has zswap enabled by default, does it?

Reply:
```
root@konata ~
❯ zcat /proc/config.gz  | grep CONFIG_ZSWAP
CONFIG_ZSWAP=y
# CONFIG_ZSWAP_DEFAULT_ON is not set
```

It doesn't.

Reply:
I see. Are you suggesting to remove this change?

Reply:
It should probably be turned into a warning, conditional on having both zram and zswap enabled.

Reply:
Hmm. Then a check of the KCONFIG and the kernel parameter. How would I do that in Nix?

Reply:
A so a conditional check on the `disable-zswap` option that I am adding in this PR then.

Reply:
No, I thought we actually had an option for zswap already. If we don't, it's probably fine to not have an explicit warning here, and maybe add a note for the future if a proper zswap module is ever added.

Reply:
I see.

Reply:
So then leave as is?

Reply:
done.